### PR TITLE
Implement asynchronous sockets

### PR DESCRIPTION
**Pull request by Mitch528**

### CODE REVIEW

File: TrueCraft.Core/World/World.cs [hunk 1]
```diff
@@ -93,8 +95,11 @@ public static World LoadWorld(string baseDirectory)
                 if (file.RootTag.Contains("Name"))
                     world.Name = file.RootTag["Name"].StringValue;
                 world.ChunkProvider = provider;
+
+                return world;
             }
-            return world;
+
+            throw new FileNotFoundException();
```
Comment (SirCmpwn) [inline TrueCraft.Core/World/World.cs]: This is not the desired behavior.

**Comment by SirCmpwn** [comment]
This is a hell of a pull request! It will require detailed testing and review before I can merge it, so expect it to take a little longer than most pull requests.

File: TrueCraft.Core/World/World.cs (hunk 1 above)
Comment (Mitch528) [inline TrueCraft.Core/World/World.cs]: If there is no world saved on the server then it won't generate one unless it catches an exception after first trying to load it. https://github.com/SirCmpwn/TrueCraft/blob/master/TrueCraft/Program.cs#L47

Comment (SirCmpwn) [inline TrueCraft.Core/World/World.cs]: I know, that's a flaw in Program.cs. Leave it be, I'll address it later.

Comment (Mitch528) [inline TrueCraft.Core/World/World.cs]: Alright.

**Comment by SirCmpwn** [comment]
Server crashes on player disconnect for failing to handle `PlayerDisconnectException`. All exceptions thrown from packet handlers should be handled gracefully - and `PlayerDisconnectException` is intentionally thrown from the disconnect handler.

**Comment by SirCmpwn** [comment]
Style convention errors here, too - private variables are `likeThis`, not `_likeThis`.

**Comment by SirCmpwn** [comment]
The exception is if you have a variable that's backing a property you need to raise `PropertyChanged` from.

File: TrueCraft.Client/MultiplayerClient.cs [hunk 2]
```diff
@@ -62,58 +80,122 @@ public void RegisterPacketHandler(byte packetId, PacketHandler handler)
...
-            new DisconnectPacket("Disconnecting").WritePacket(Stream);
-            Stream.BaseStream.Flush();
+            if (!Connected)
+                return;
+
+            Interlocked.CompareExchange(ref _connected, 0, 1);
+
+            QueuePacket(new DisconnectPacket("Disconnecting"));
+
+            Client.Client.Shutdown(SocketShutdown.Send);
```
Comment (SirCmpwn) [inline TrueCraft.Client/MultiplayerClient.cs]: This is nonblocking. Make sure that the disconnect packet is sent before Disconnect returns.

File: TrueCraft/MultiplayerServer.cs [hunk 3]
```diff
@@ -61,20 +62,20 @@ public bool BlockUpdatesEnabled
...
-        private Thread NetworkWorker;
         private TcpListener Listener;
         private readonly PacketHandler[] PacketHandlers;
         private IList<ILogProvider> LogProviders;
         internal object ClientLock = new object();
-        private bool ShuttingDown = false;
+        
         private QueryProtocol QueryProtocol;
 
+        public bool ShuttingDown { get; private set; }
```
Comment (SirCmpwn) [inline TrueCraft/MultiplayerServer.cs]: This should be internal, not public.

**Comment by Mitch528** [comment]
Done.

**Comment by SirCmpwn** [comment]
:tada:

**Pull request merged**